src/orange/orangeinp/Solid.hh

```diff
@@ -0,0 +1,185 @@
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/orangeinp/Solid.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+#include "corecel/math/Turn.hh"
+
+#include "ConvexRegion.hh"
+#include "ObjectInterface.hh"
+
+namespace celeritas
+{
+namespace orangeinp
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Define the angular region of a solid.
+ *
+ * This angle is a cross section perpendicular to the z axis, with a start
+ * angle of zero corresponding to the positive x axis. An interior angle of one
+ * results in no radial excluded in the resulting solid. A interior angle of
+ * more than 0.5 turns (180 degrees) results in a wedge being subtracted from
+ * the solid, and an angle of less than or equal to 0.5 turns results in a
+ * solid
+ */
+class SolidEnclosedAngle
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using SenseWedge = std::pair<Sense, InfWedge>;
+    //!@}
+
+  public:
+    //! Default to "all angles"
+    SolidEnclosedAngle() = default;
+
+    // Construct from a starting angle and interior angle
+    SolidEnclosedAngle(Turn start, Turn interior);
+
+    // Construct a wedge shape to intersect (inside) or subtract (outside)
+    SenseWedge make_wedge() const;
+
+    // Whether the enclosed angle is not a full circle
+    explicit inline operator bool() const;
+
+    //! Starting angle
+    Turn start() const { return start_; }
+
+    //! Interior angle
+    Turn interior() const { return interior_; }
+
+  private:
+    Turn start_{0};
+    Turn interior_{1};
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * A hollow shape with an optional start and end angle.
+ */
+class SolidBase : public ObjectInterface
+{
+  public:
+    // Construct a volume from this object
+    NodeId build(VolumeBuilder&) const final;
+
+    // Write the shape to JSON
+    void output(JsonPimpl*) const final;
+
+    //! Interior convex region interface for construction and access
+    virtual ConvexRegionInterface const& interior() const = 0;
+
+    //! Optional excluded
+    virtual ConvexRegionInterface const* excluded() const = 0;
+
+    //! Angular restriction to add
+    virtual SolidEnclosedAngle const& enclosed_angle() const = 0;
+
+  protected:
+    //!@{
+    //! Allow construction and assignment only through daughter classes
+    SolidBase() = default;
+    virtual ~SolidBase() = default;
+    CELER_DEFAULT_COPY_MOVE(SolidBase);
+    //!@}
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * A shape that is hollow, is truncated azimuthally, or both.
```

Comment:
I would say "interior excluded region" and/or a wedge subtracted.

I also don't understand why it couldn't be neither, though I see that that is how it is coded.

Reply:
I'm deliberately trying to prevent two `Object` types from being the same representation unless they compare equal... maybe I'll relax that eventually, but I think for now I don't think it's a problem to keep the restriction. Or do you think it's better to have a "Solid" that's the same as a "Shape", and later if we need to we can add a "simplify" method that'll let solids decay to shapes? That *would* remove some logic from the Geant4 translator...

Reply:
So if a geometry consists of only simple cuboids for example (with no internal regions or wedges) then solids are never created? In a sense it would be nice to have geometry elements go through the same levels of abstraction, but it might not be worth it if that adds a bunch more code that is essentially a no-op. To weight in further I think need so study the forthcoming design diagram/document, because I am not clear on what exactly happens on each level of abstraction.

Reply:
just `Shape`s, not `Solid`. The abstraction is that they're both `Object` interfaces. So in Geant4 a "cone solid" could produce either a `ConeShape` in its simplest form or a `ConeSolid` if it's truncated.

Reply:
Okay I see. The nomenclature here is difficult especially since Geant4 nomenclature clashes with ORANGE nomenclature (recognizing that may be unavoidable).

Reply:
Well, I've tried to delineate it a bit: "shape" is going to be for the simpler objects, "solids" are the weird Geant4 ones. See https://github.com/celeritas-project/celeritas/issues/1078#issue-2064118326